core/domain/exp_domain.py

```diff
@@ -3129,6 +3108,41 @@ def _convert_states_v17_dict_to_v18_dict(cls, states_dict):
 
         return states_dict
 
+    @classmethod
+    def _convert_states_v18_dict_to_v19_dict(cls, states_dict):
+        """Converts from version 18 to 19. Version 19 adds training_data
+        parameter to each answer group to store training data of that
+        answer group.
+
+        Args:
+            states_dict: dict. A dict where each key-value pair represents,
+                respectively, a state name and a dict used to initialize a
+                State domain object.
+
+        Returns:
+            dict. The converted states_dict.
+        """
+        for state_dict in states_dict.values():
+            for answer_group in state_dict['interaction']['answer_groups']:
+                if answer_group['rule_specs']:
+                    training_data = []
+                    classifier_rule = None
+                    for rule in answer_group['rule_specs']:
+                        if rule['rule_type'] == RULE_TYPE_CLASSIFIER:
+                            training_data = rule['inputs']['training_data']
+                            classifier_rule = rule
+                            break
+                    if classifier_rule:
+                        answer_group['rule_specs'].remove(classifier_rule)
+
+                    answer_group['training_data'] = training_data
+
+                    if training_data == [] and answer_group['rule_specs'] == []:
+                        state_dict['interaction']['answer_groups'].remove(
```

Comment:
This one is even iffier because you are currently iterating through the list that you're mutating. Better to store the indexes somewhere and then remove them all at the end.

Reply:
Done